### PR TITLE
feat(sdk): Filter Dropdown

### PR DESCRIPTION
**Pull request by oisincoveney**
Adds the Filter dropdown to the interactive question picker. Here's a demo:

https://github.com/user-attachments/assets/0121dcbd-2120-4386-a6e3-a5efd5642ae9

### CODE REVIEW

File: enterprise/frontend/src/embedding-sdk/components/private/InteractiveQuestion/components/Filter/Filter.stories.tsx [hunk 1]
```diff
@@ -0,0 +1,34 @@
+import { InteractiveQuestion } from "embedding-sdk";
+import { CommonSdkStoryWrapper } from "embedding-sdk/test/CommonSdkStoryWrapper";
+import { Box, Flex } from "metabase/ui";
+
+import { Filter } from "./Filter";
+
+const QUESTION_ID = (window as any).QUESTION_ID || 12;
+
+export default {
+  title: "EmbeddingSDK/InteractiveQuestion/Filter",
```
Comment (heypoom) [inline enterprise/frontend/src/embedding-sdk/components/private/InteractiveQuestion/components/Filter/Filter.stories.tsx:10]: Just to confirm, would this be named `FilterDropdown` or `Filter`?

File: enterprise/frontend/src/embedding-sdk/components/private/InteractiveQuestion/components/Filter/Filter.tsx [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+
+  const filters = useMemo(() => getFilterItems(query), [query]);
+
+  const [step, setStep] = useState<"list" | "picker">("list");
+
+  const [opened, { close, toggle }] = useDisclosure(false, {
+    onOpen: () => {
+      setStep(filters.length === 0 ? "picker" : "list");
+    },
+  });
```
Comment (heypoom) [inline enterprise/frontend/src/embedding-sdk/components/private/InteractiveQuestion/components/Filter/Filter.tsx:80]: Just an observation that I saw this block across multiple components while reviewing this stack of PRs, I'm thinking if we can generalize the notion of "dropdown" across these components.

File: enterprise/frontend/src/embedding-sdk/components/private/InteractiveQuestion/components/Filter/Filter.tsx [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+      onUpdateFilter(selectedFilter.filter, nextFilter);
+    } else {
+      onAddFilter(nextFilter);
+    }
+  };
+
+  const label = match(filters.length)
+    .with(0, () => "Filter")
+    .with(1, () => "1 filter")
+    .otherwise(value => `${value} filters`);
```
Comment (heypoom) [inline enterprise/frontend/src/embedding-sdk/components/private/InteractiveQuestion/components/Filter/Filter.tsx:98]: do we need to localize this?

File: enterprise/frontend/src/embedding-sdk/components/private/InteractiveQuestion/components/Filter/Filter.tsx [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+};
+
+export const Filter = ({
+  withColumnItemIcon,
+  withColumnGroupIcon,
+  withCustomExpression,
+}: Pick<
+  FilterColumnPickerProps,
+  "withColumnItemIcon" | "withColumnGroupIcon" | "withCustomExpression"
+>) => {
```
Comment (heypoom) [inline enterprise/frontend/src/embedding-sdk/components/private/InteractiveQuestion/components/Filter/Filter.tsx:166]: Should we extract this Pick<...> block to a separate type, so we can use it in both components?

**Review by heypoom — APPROVED** [review]
I love the UX of the dropdown! Chill mode is so cool

File: enterprise/frontend/src/embedding-sdk/components/private/InteractiveQuestion/components/Filter/Filter.stories.tsx (hunk 1 above)
Comment (oisincoveney) [reply enterprise/frontend/src/embedding-sdk/components/private/InteractiveQuestion/components/Filter/Filter.stories.tsx:10]: that should be sorted now 😄

File: enterprise/frontend/src/embedding-sdk/components/private/InteractiveQuestion/components/Filter/Filter.tsx (hunk 2 above)
Comment (oisincoveney) [reply enterprise/frontend/src/embedding-sdk/components/private/InteractiveQuestion/components/Filter/Filter.tsx:80]: It's definitely something I want to do in the future once all of this is done - I want to make it work before it has to work well 🙂

File: enterprise/frontend/src/embedding-sdk/components/private/InteractiveQuestion/components/Filter/Filter.tsx (hunk 3 above)
Comment (oisincoveney) [reply enterprise/frontend/src/embedding-sdk/components/private/InteractiveQuestion/components/Filter/Filter.tsx:98]: yeah we should do that

**Comment by WiNloSt** [comment]
One quick thing I noticed is that the filter popover has bold texts, but not the inline version

#### Filter popover
image

#### (just) Filter
image


These screenshots are captured from provided storybook stories.

**Comment by oisincoveney** [comment]
> One quick thing I noticed is that the filter popover has bold texts, but not the inline version
> 
> 
> 
> #### Filter popover
> 
> image
> 
> 
> 
> #### (just) Filter
> 
> image
> 
> 
> 
> 
> 
> These screenshots are captured from provided storybook stories.

Weird. I'll take a look

File: enterprise/frontend/src/embedding-sdk/components/private/InteractiveQuestion/components/Filter/FilterPicker/FilterPicker.stories.tsx
Comment (WiNloSt) [inline enterprise/frontend/src/embedding-sdk/components/private/InteractiveQuestion/components/Filter/FilterPicker/FilterPicker.stories.tsx:1]: I don't think this file should be here, or named `FilterPicker` since it uses the `Filter` component which renders `FilterPicker` inside. I was actually confused by this at first.

It should just be named `Filter.stories.tsx` and lives one level up.

File: enterprise/frontend/src/embedding-sdk/components/private/InteractiveQuestion/components/index.ts [hunk 5]
```diff
@@ -3,9 +3,9 @@ export * from "./Breakout";
 export * from "./ChartTypeSelector";
 export * from "./ChartTypeSelectorList";
 export * from "./Filter";
-export * from "./FilterBar";
-export * from "./FilterButton";
-export * from "./FilterPicker";
+export * from "./Filter/FilterBar";
+export * from "./Filter/FilterButton";
+export * from "./Filter/FilterPicker/FilterPicker";
```
Comment (WiNloSt) [inline enterprise/frontend/src/embedding-sdk/components/private/InteractiveQuestion/components/index.ts:8]:
```suggestion
export * from "./Filter/FilterPicker";
```

File: enterprise/frontend/src/embedding-sdk/components/private/InteractiveQuestion/components/util/ToolbarButton/ToolbarButton.tsx [hunk 6]
```diff
@@ -12,16 +12,16 @@ function _ToolbarButton(
     ...buttonProps
   }: {
     label: ReactNode;
-    icon: IconName;
-    isHighlighted: boolean;
+    icon?: IconName;
+    isHighlighted?: boolean;
```
Comment (WiNloSt) [inline enterprise/frontend/src/embedding-sdk/components/private/InteractiveQuestion/components/util/ToolbarButton/ToolbarButton.tsx:16]: I've checked that we always pass these 2 props, also with reverting to make them required still pass `yarn type-check` so I suggest reverting it.

```suggestion
    icon: IconName;
    isHighlighted: boolean;
```

This will allows us to revert the change below as well.

File: enterprise/frontend/src/embedding-sdk/components/private/InteractiveQuestion/components/Filter/FilterDropdown/FilterDropdown.tsx [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+
+  const onRemoveItem = (item: SDKFilterItem) => {
+    item.onRemoveFilter();
+    if (filters.length === 1) {
+      close();
+    }
+  };
+
+  return (
+    <MultiStepPopover currentStep={step} opened={opened} onClose={close}>
```
Comment (WiNloSt) [inline enterprise/frontend/src/embedding-sdk/components/private/InteractiveQuestion/components/Filter/FilterDropdown/FilterDropdown.tsx:59]: I find controlling the multi-step popover state with 2 props hard to reason about.

If you think about the state of the popover it can only be one of:
- closed
- open: stage 1
- open: stage 2
- open: stage n

This way we don't need to control these 2 states individually and try to make them sync to make sure `MultiStepPopover` works as expected.

The parent component can only focus on a single state machine that transitions the state of the multi-step popover instead of 2.

We can compute `open` prop in a number of ways.
1. provide `closeStep` name
2. `null` to denote the closed state
3. infer that from the step name that doesn't match any of the `MultiStepPopover.Step` value.

I like 1 the most since it's simple, we could add a typedoc to explain these details. But anything works.

File: enterprise/frontend/src/embedding-sdk/components/private/InteractiveQuestion/components/Filter/FilterDropdown/FilterDropdown.tsx [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+  if (!question) {
+    return null;
+  }
+
+  const onQueryChange = (query: Lib.Query) => {
+    updateQuestion(question.setQuery(query), { run: true });
+  };
+
+  return (
+    <FilterDropdownInner
```
Comment (WiNloSt) [inline enterprise/frontend/src/embedding-sdk/components/private/InteractiveQuestion/components/Filter/FilterDropdown/FilterDropdown.tsx:104]: This component only accepts 2 props, so the rest wouldn't do anything, so we should remove them.

https://github.com/metabase/metabase/blob/61de2e5c1651a058ab7f783e010878dfcc77d242/enterprise/frontend/src/embedding-sdk/components/private/InteractiveQuestion/components/Filter/FilterDropdown/FilterDropdown.tsx#L23-L26

File: enterprise/frontend/src/embedding-sdk/components/private/InteractiveQuestion/components/Filter/FilterBadgeList.tsx
Comment (WiNloSt) [inline enterprise/frontend/src/embedding-sdk/components/private/InteractiveQuestion/components/Filter/FilterBadgeList.tsx:1]: I wonder if there's a way to rename some of these components to make it easier to understand?

e.g. `BadgeList` and `FilterBar` are almost the same component, but they're controlled differently by either a dropdown filter or just a filter. But with their names not resembling each other at all, it's hard to understand why these 2 components look so similar.

I understand that the reason is mostly because the popover/dropdown targets are different based on whether or not a filter is in a dropdown, but I'd still expect some kind of name that makes sense.

File: enterprise/frontend/src/embedding-sdk/components/private/InteractiveQuestion/components/Filter/hooks/use-filter-handlers.ts [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+      stageIndex,
+      currentFilterClause,
+      nextFilterClause,
+    );
+    onQueryChange(nextQuery);
+  };
+
+  return {
+    getFilterName,
+    onAddFilter,
```
Comment (WiNloSt) [inline enterprise/frontend/src/embedding-sdk/components/private/InteractiveQuestion/components/Filter/hooks/use-filter-handlers.ts:38]: we only seem to use `onAddFilter`, do we even need the rest 3 handlers?

**Review by WiNloSt — APPROVED** [review]
I've tested locally and the component works really well 👍

File: enterprise/frontend/src/embedding-sdk/components/private/InteractiveQuestion/components/util/ToolbarButton/ToolbarButton.tsx (hunk 6 above)
Comment (oisincoveney) [reply enterprise/frontend/src/embedding-sdk/components/private/InteractiveQuestion/components/util/ToolbarButton/ToolbarButton.tsx:16]: I'm actually going to use this component in the next two PRs without `icon` since I want the `Save` button to use the same styles. I _could_ make a new component but I figured it would be easier to reuse this in the future. 

Would you prefer that I change it here, and then make the change in the `Save` button PR?

Comment (WiNloSt) [reply enterprise/frontend/src/embedding-sdk/components/private/InteractiveQuestion/components/util/ToolbarButton/ToolbarButton.tsx:16]: That will make the change easier to understand for each PR, so I prefer that yeah. It's hard to understand a PR with outside context.

File: enterprise/frontend/src/embedding-sdk/components/private/InteractiveQuestion/components/Filter/FilterBadgeList.tsx
Comment (oisincoveney) [reply enterprise/frontend/src/embedding-sdk/components/private/InteractiveQuestion/components/Filter/FilterBadgeList.tsx:1]: Yeah, I think this is more of a consequence of the other two components having `{name}BadgeList` so I kept with the consistency, but it doesn't make sense to have both. I'll take a look and see what the best thing to do here is

File: enterprise/frontend/src/embedding-sdk/components/private/InteractiveQuestion/components/Filter/FilterDropdown/FilterDropdown.tsx (hunk 7 above)
Comment (oisincoveney) [reply enterprise/frontend/src/embedding-sdk/components/private/InteractiveQuestion/components/Filter/FilterDropdown/FilterDropdown.tsx:59]: after a conversation we've decided to use `null` as the `closed` state.

File: enterprise/frontend/src/embedding-sdk/components/private/InteractiveQuestion/components/Filter/FilterBadgeList.tsx
Comment (oisincoveney) [reply enterprise/frontend/src/embedding-sdk/components/private/InteractiveQuestion/components/Filter/FilterBadgeList.tsx:1]: this one is really annoying to solve actually. I'm not totally sure what the best course of action would be considering the structure of the Mantine popover
